### DISCUSSION TITLE
Difficulties sorfted alphabetically, why?

### DISCUSSION THREAD

**Discussion by KoleckOLP** [ppy/osu#19876] [Q&A]
when I search for a map difficulties are sorted alphabetically not by difficulty why?
![image](https://user-images.githubusercontent.com/22729896/185765291-e26788ac-b925-435e-a88e-aa61059117f9.png)

it does not seem to be only if I search just some maps are sorted bu difficulty and some by alphabet,
could they be ordered by star rating just like before?
is any way to fix this?

**Comment by smoogipoo** [comment]
Does it happen after reloading song select? I believe this was because the difficulties weren't calculated...

**Reply by KoleckOLP** [reply]
They were fine before, I updated osu and now it's like this.
And I am afraid of reloading all my songs because I have way too many. Should I?

**Reply by peppy** [reply]
No need to do anything like that yet.

If possible, could you please attach your `client.realm` file?

And can you confirm that this happens when loading song select, before searching and 100% of the time? At a code level I can't immediately see how this can happen, even in the case of difficulty calculation failing.

**Reply by smoogipoo** [reply]
Based on the screenshot, there's two observations:
1. The group icon on the beatmap set is gray, indicating SR=0.
2. `CarouselBeatmap`s are only sorted by ruleset and then SR: https://github.com/ppy/osu/blob/594353fca18efb3ef4c9fbbf03459e6188243bcc/osu.Game/Screens/Select/Carousel/CarouselBeatmap.cs#L87-L102

Given ruleset (0) and SR (0) are matching, `CarouselBeatmapSet`/`CarouselGroup` falls back to a by-id match: https://github.com/ppy/osu/blob/594353fca18efb3ef4c9fbbf03459e6188243bcc/osu.Game/Screens/Select/Carousel/CarouselGroup.cs#L89-L96 which probably just so happens to match title-sorting in this case - whether it's because that's the order in which they're read from the archive at import time or otherwise.

**Reply by KoleckOLP** [reply]
here is the client.realm it's too big to send over github (I hope it's the right file.)
https://fosfacz-my.sharepoint.com/:u:/g/personal/kurz_fosfa_cz/ESfCFnZ6UhlDr7bvJaWVdv0BrKc_mKMkk9lVOIV0WKYtkQ?e=6dS9qU

it's seems that more songs have this issue
I don't remembere having the issue in build 2022.719, but I do have the issue in 2022.816 and 2022.821 (didn't try any other builds)

![image](https://user-images.githubusercontent.com/22729896/185947516-c67af7a9-b343-4a54-85ca-942105c69c16.png)


the problem happens even whithout searching, but seem to only happen on maps that are not ranked, ranked maps seem to be fine.

**Reply by peppy** [reply]
Thanks for the database. For good measure can you also attach your logs after visiting one of the beatmaps at song select which shows the issue? As smoogipoo mentioned, it definitely looks like difficulty calculation is not working.

**Reply by KoleckOLP** [reply]
I tried to get logs for the issue, but the issue is gone.
all the jump training maps that had difficulties listed alphabetically are now ordered by star rating as it was before.
I don't know what to say.

here is a log anyway.
[logs.zip](https://github.com/ppy/osu/files/9399283/logs.zip)

**Reply by peppy** [reply]
Looking at your database, it seems highly likely that background difficulty calculation was still running when you were seeing the issue. The number of beatmaps you have loaded would contribute to this.

We do want to expose this better in the future so it's obvious what is happening, and also improve the performance so it doesn't take so long to complete processing (right now the reprocess operation is doing more than just difficulty recalculation, which would be adding to the time it takes).

**Reply by KoleckOLP** [reply]
But isn't it strange that in 2022.719 I didn't had the issue and updating to 2022.816 I did.
Was there some difficulty calculation update between these updates?

**Reply by peppy** [reply]
Yes, we introduced the concept of versioning for difficulty calculation so we can trigger recalculation in the future. As part of this, all local star ratings were reset. See https://github.com/ppy/osu/pull/19272.
